// @gathertown/gather-game-common/src/public/gameMap.ts
export {
  AreaCategory,
  coordsMapToPositions,
  convertAreaCoordsToCoordsMap,
  convertCoordsToBBoxes,
  convertDBDesksCoordsToCoordsMap,
  convertDBDesksToDesks,
  convertDesksToDBDesks,
  convertNookInfoToOccupiedNookTilesMap,
  convertMapObjectToWireObject,
  convertStringToEnabledChat,
  convertStringToPresetTutorialGroupSetId,
  convertTutorialTasksToWireTutorialTasks,
  convertWireAreasToAreas,
  convertWireAreasToDBAreas,
  convertWireObjectToMapObjectPartial,
  convertWireObjectToMapObjectDBPartial,
  convertWireTutorialTasksToTutorialTasks,
  desksToNooks,
  EnabledChat,
  findMissingMapObjectFields,
  FloorBlock,
  generateEmptyDeskCoordsMap,
  Interaction,
  interactionOpensModal,
  isMapObject,
  isMapObjectDB,
  isNook,
  MAP_MAX_DIMENSION,
  mergeNooks,
  nookToDeprecatedSpaces,
  objectsArrayToDict,
  objectsDictToArray,
  deprecatedSpacesToNooks,
  PresetTutorialGroupSetIds,
  TileType,
  isTileType,
  WallBlock,
  WallFloorDB,
} from "../GameMap";

export type {
  Area,
  Areas,
  CoordsMap,
  BaseArea,
  DBArea,
  DBAreas,
  DBDesks,
  Desks,
  SpawnTile,
  PortalTile,
  TileEffect,
  GameMapBaseDoNotUseDirectly,
  GameMapDB,
  GameMapHttpPublic,
  MapNameAndId,
  MapObject,
  MapObjectDB,
  MapObjectToAdd,
  ModalInteraction,
  Nook,
  NookInfo,
  NookTile,
  OccupiedNookTilesMap,
  Tile,
  TutorialTasks,
  WallFloorBlock,
} from "../GameMap";

// We reexport some of the types from the protobuf `events.ts` because this is a more likely and logical place people
// would search for them. Also it makes it possible to auto-import with Intellisense, since it's not possible from
// public/events.ts.
export { Space, Portal, SpawnPoint } from "../generated_DO_NOT_TOUCH/events";
