// @gathertown/gather-game-common/dist/src/playerUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VehicleAction = exports.whisperIdToColor = exports.directionToFacePlayer = exports.nextSpriteDirection = exports.positionAfterMove = exports.oppositeMoveDirection = exports.getMoveDirFromSpriteDir = exports.DEFAULT_NAME = exports.generateDefaultPlayer = void 0;
const positionUtils_1 = require("./positionUtils");
const Player_1 = require("./Player");
const generateDefaultPlayer = (playerId) => {
    const player = new Player_1.Player(playerId);
    player.x = -99999;
    player.y = -99999;
    player.emote = "";
    player.workCondition = 0;
    player.away = true;
    player.lastActive = new Date().toISOString();
    player.connected = true;
    return player;
};
exports.generateDefaultPlayer = generateDefaultPlayer;
exports.DEFAULT_NAME = "Anonymous";
function getMoveDirFromSpriteDir(direction) {
    switch (direction) {
        case Player_1.SpriteDirection.Dance1:
        case Player_1.SpriteDirection.Dance2:
        case Player_1.SpriteDirection.Dance3:
        case Player_1.SpriteDirection.Dance4:
            return Player_1.MoveDirection.Dance;
        case Player_1.SpriteDirection.Down:
        case Player_1.SpriteDirection.DownAlt:
            return Player_1.MoveDirection.Down;
        case Player_1.SpriteDirection.Up:
        case Player_1.SpriteDirection.UpAlt:
            return Player_1.MoveDirection.Up;
        case Player_1.SpriteDirection.Left:
        case Player_1.SpriteDirection.LeftAlt:
            return Player_1.MoveDirection.Left;
        case Player_1.SpriteDirection.Right:
        case Player_1.SpriteDirection.RightAlt:
            return Player_1.MoveDirection.Right;
        default:
            return null;
    }
}
exports.getMoveDirFromSpriteDir = getMoveDirFromSpriteDir;
function oppositeMoveDirection(direction) {
    switch (direction) {
        case Player_1.MoveDirection.Left:
            return Player_1.MoveDirection.Right;
        case Player_1.MoveDirection.Right:
            return Player_1.MoveDirection.Left;
        case Player_1.MoveDirection.Up:
            return Player_1.MoveDirection.Down;
        case Player_1.MoveDirection.Down:
            return Player_1.MoveDirection.Up;
        default:
            return null;
    }
}
exports.oppositeMoveDirection = oppositeMoveDirection;
function positionAfterMove(start, dir, dist = 1) {
    const { x, y, map } = start;
    switch (dir) {
        case Player_1.MoveDirection.Left:
            return { x: x - dist, y, map };
        case Player_1.MoveDirection.Right:
            return { x: x + dist, y, map };
        case Player_1.MoveDirection.Up:
            return { x, y: y - dist, map };
        case Player_1.MoveDirection.Down:
            return { x, y: y + dist, map };
        default:
            return start;
    }
}
exports.positionAfterMove = positionAfterMove;
function nextSpriteDirection(current, direction) {
    if (direction === Player_1.MoveDirection.Left && current === Player_1.SpriteDirection.Left) {
        return Player_1.SpriteDirection.LeftAlt;
    }
    else if (direction === Player_1.MoveDirection.Right && current === Player_1.SpriteDirection.Right) {
        return Player_1.SpriteDirection.RightAlt;
    }
    else if (direction === Player_1.MoveDirection.Up && current === Player_1.SpriteDirection.Up) {
        return Player_1.SpriteDirection.UpAlt;
    }
    else if (direction === Player_1.MoveDirection.Down && current === Player_1.SpriteDirection.Down) {
        return Player_1.SpriteDirection.DownAlt;
    }
    else if (direction === Player_1.MoveDirection.Dance && current === Player_1.SpriteDirection.Dance1) {
        return Player_1.SpriteDirection.Dance2;
    }
    else if (direction === Player_1.MoveDirection.Dance && current === Player_1.SpriteDirection.Dance2) {
        return Player_1.SpriteDirection.Dance3;
    }
    else if (direction === Player_1.MoveDirection.Dance && current === Player_1.SpriteDirection.Dance3) {
        return Player_1.SpriteDirection.Dance4;
    }
    else if (direction === Player_1.MoveDirection.Left) {
        return Player_1.SpriteDirection.Left;
    }
    else if (direction === Player_1.MoveDirection.Right) {
        return Player_1.SpriteDirection.Right;
    }
    else if (direction === Player_1.MoveDirection.Up) {
        return Player_1.SpriteDirection.Up;
    }
    else if (direction === Player_1.MoveDirection.Down) {
        return Player_1.SpriteDirection.Down;
    }
    else if (direction === Player_1.MoveDirection.Dance) {
        return Player_1.SpriteDirection.Dance1;
    }
    return current;
}
exports.nextSpriteDirection = nextSpriteDirection;
function directionToFacePlayer(player, target) {
    if (player.map !== target.map)
        return null;
    if ((0, positionUtils_1.manhattanDistance)((0, positionUtils_1.dangerouslyCastToPoint)(player), (0, positionUtils_1.dangerouslyCastToPoint)(target)) !== 1)
        return null;
    if (player.x - target.x === 1) {
        return Player_1.MoveDirection.Left;
    }
    else if (player.x - target.x === -1) {
        return Player_1.MoveDirection.Right;
    }
    else if (player.y - target.y === 1) {
        return Player_1.MoveDirection.Up;
    }
    else if (player.y - target.y === -1) {
        return Player_1.MoveDirection.Down;
    }
    return null;
}
exports.directionToFacePlayer = directionToFacePlayer;
function whisperIdToColor(id) {
    if (id.length === 7)
        return id;
    return "#" + id.slice(0, 6);
}
exports.whisperIdToColor = whisperIdToColor;
var VehicleAction;
(function (VehicleAction) {
    VehicleAction["Mount"] = "mount";
    VehicleAction["Dismount"] = "dismount";
})(VehicleAction = exports.VehicleAction || (exports.VehicleAction = {}));
